### ISSUE TITLE
CircuitOperation has non-deterministic diagrams, making tests hard to write

### ISSUE THREAD

**Issue by Strilanc** [quantumlib/Cirq#4240]
I filled in a diagram test from output during one run:

```
    cirq.testing.assert_has_diagram(
        cirq.Circuit(
            cirq.Moment(
                cirq.CircuitOperation(
                    cirq.FrozenCircuit(cirq.X(cirq.LineQubit(0)))),
            ),
        ),
        """
0: ───Circuit_0x92933a3050ff70c4:───
      [ 0: ───X───              ]
        """)
```

But then it fails in the next run:

```
E       Diagram of actual circuit:
E       0: ───Circuit_0xd4e3300492f7877d:───
E             [ 0: ───X───              ]
E       
E       Desired text diagram:
E       0: ───Circuit_0x92933a3050ff70c4:───
E             [ 0: ───X───              ]
E       
E       Highlighted differences:
E       0: ───Circuit_0x███33█████f█████:───
E             [ 0: ───X───              ]
```

The issue is that the hash id is going into the diagram, but hash ids are not stable over time.

The hash is entirely based on the contents, which we can see anyways. I suggest just dropping it.

**Comment by 95-martin-orion** [comment]
My reasoning behind including the hash:
- It's easier to visually compare hash values than large CircuitOperations
- Tests can embed `hash(circuitop)` of `circuit.diagram_name()` in the diagram string if necessary

An example from the CircuitOperation tests:
https://github.com/quantumlib/Cirq/blob/099fc740c53a3e54b2bece568f246c7324fd3e11/cirq-core/cirq/circuits/circuit_operation_test.py#L350-L361

**Comment by thanacles** [comment]
Is https://github.com/quantumlib/Cirq/pull/4712 what you are thinking of?  I (in my ignorance) feel like CircuitOperation could benefit from:
1. Being a Gate instead of an Operation (CircuitGate?)
2. Getting a similar treatment as this: https://github.com/quantumlib/Cirq/pull/4201

**Comment by 95-martin-orion** [comment]
Responded to (2) in #3605.

Your thoughts on CircuitGate mirror the early design stages for CircuitOperation - we strongly considered that option, but ultimately determined that CircuitOperation was the cleaner solution. This mostly boiled down to Circuits being more similar to Operations (both have defined qubits, while Gates do not) and Operations having better access to certain features (e.g. tags are defined at the Operation level).

**Comment by thanacles** [comment]
I believe this issue is now fixed with merge of PR #4712
